### ISSUE TITLE
Unhandled exception on mutex lock in Synchronizer in log appender callback

### ISSUE THREAD

**Issue by mbautin** [YugaByte/yugabyte-db#239]
```
Program terminated with signal 6, Aborted.
#0  0x00002b640e15d067 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb)
(gdb)
(gdb) where
#0  0x00002b640e15d067 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00002b640e15e46a in __GI_abort () at abort.c:89
#2  0x00002b640d6c45bd in __gnu_cxx::__verbose_terminate_handler() () from /home/yugabyte/tserver/bin/../lib/linuxbrew/libstdc++.so.6
#3  0x00002b640d6c2626 in ?? () from /home/yugabyte/tserver/bin/../lib/linuxbrew/libstdc++.so.6
#4  0x00002b640d6c2671 in std::terminate() () from /home/yugabyte/tserver/bin/../lib/linuxbrew/libstdc++.so.6
#5  0x00002b640d6c2888 in __cxa_throw () from /home/yugabyte/tserver/bin/../lib/linuxbrew/libstdc++.so.6
#6  0x00002b640d6ea4ff in std::__throw_system_error(int) () from /home/yugabyte/tserver/bin/../lib/linuxbrew/libstdc++.so.6
#7  0x00002b6403e175c4 in lock (this=0x2b6aeedd8060) at /n/jenkins/linuxbrew/linuxbrew_2018-03-16T16_38_10/Cellar/gcc/5.5.0/include/c++/5.5.0/mutex:139
#8  lock_guard (__m=..., this=<synthetic pointer>) at /n/jenkins/linuxbrew/linuxbrew_2018-03-16T16_38_10/Cellar/gcc/5.5.0/include/c++/5.5.0/mutex:386
#9  yb::Synchronizer::StatusCB (this=0x2b6aeedd8060, status=...) at ../../../../../src/yb/util/async_util.h:63
#10 0x00002b6403de617c in Run (a1=..., this=0x112901f430) at ../../../../../src/yb/gutil/callback.h:451
#11 yb::consensus::PeerMessageQueue::LocalPeerAppendFinished(yb::OpIdPB const&, yb::Callback<void (yb::Status const&)> const&, yb::Status const&) (this=0x11ee090900, id=..., callback=..., status=...) at ../../../../../src/yb/consensus/consensus_queue.cc:275
#12 0x00002b6403df4986 in Run (a1=..., this=0x12ff12f8) at ../../../../../src/yb/gutil/callback.h:451
#13 yb::consensus::LogCache::LogCallback(long, bool, yb::Callback<void (yb::Status const&)> const&, yb::Status const&) (this=<optimized out>, last_idx_in_batch=<optimized out>, borrowed_memory=<optimized out>, user_callback=..., log_status=...) at ../../../../../src/yb/consensus/log_cache.cc:240
#14 0x00002b640406ebb2 in Run (a1=..., this=<optimized out>) at ../../../../../src/yb/gutil/callback.h:451
#15 yb::log::Log::Appender::GroupWork (this=0x670f028a0) at ../../../../../src/yb/consensus/log.cc:274
#16 0x00002b640407305d in yb::log::Log::Appender::ProcessBatch (this=0x670f028a0, entry_batch=0x0) at ../../../../../src/yb/consensus/log.cc:203
#17 0x00002b640407837f in operator() (__args#0=0x0, this=0x3fceb890) at /n/jenkins/linuxbrew/linuxbrew_2018-03-16T16_38_10/Cellar/gcc/5.5.0/include/c++/5.5.0/functional:2267
#18 ProcessItem (item=0x0, this=0x3fceb760) at ../../../../../src/yb/util/taskstream.h:208
#19 yb::TaskStreamImpl<yb::log::LogEntryBatch>::Run (this=0x3fceb760) at ../../../../../src/yb/util/taskstream.h:185
#20 0x00002b6409a19f74 in operator() (this=<optimized out>) at /n/jenkins/linuxbrew/linuxbrew_2018-03-16T16_38_10/Cellar/gcc/5.5.0/include/c++/5.5.0/functional:2267
#21 Run (this=<optimized out>) at ../../../../../src/yb/util/threadpool.h:444
#22 yb::ThreadPool::DispatchThread (this=0x27cb200, permanent=false) at ../../../../../src/yb/util/threadpool.cc:608
#23 0x00002b6409a166e6 in operator() (this=0x4dc135e38) at /n/jenkins/linuxbrew/linuxbrew_2018-03-16T16_38_10/Cellar/gcc/5.5.0/include/c++/5.5.0/functional:2267
#24 yb::Thread::SuperviseThread (arg=<optimized out>) at ../../../../../src/yb/util/thread.cc:602
#25 0x00002b640df15694 in start_thread (arg=0x2b6ae16dd700) at pthread_create.c:333
#26 0x00002b640e2103cd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

**Comment by mbautin** [comment]
Here is the mutex object from gdb:
```
(gdb) print this
$1 = (std::mutex * const) 0x2b6aeedd8060
(gdb) print *this
$2 = {<std::__mutex_base> = {_M_mutex = {__data = {__lock = -287473536, __count = 11114, __owner = 164156713, __nusers = 11108, __kind = 744961680, __spins = 2, __elision = 0, __list = {__prev = 0x6e15f4600, __next = 0x106ea6af00}},
      __size = "\200\200\335\356j+\000\000)\325\310\td+\000\000\220\066g,\002\000\000\000\000F_\341\006\000\000\000\000\257\246n\020\000\000", __align = 47738274021504}}, <No data fields>}
```

The mutex state does not seem valid. It looks like the Synchronizer object has gone out of scope and has become invalid.

**Comment by mbautin** [comment]
My previous theory for why SnoozeFailureDetectorUnlocked might fail is probably incorrect, because we only transition the tablet peer to the running state after starting the consensus:

```c++
Status TabletPeer::Start(const ConsensusBootstrapInfo& bootstrap_info) {
  {
    std::lock_guard<simple_spinlock> l(state_change_lock_);
    TRACE("Starting consensus");

    VLOG(2) << "T " << tablet_id() << " P " << consensus_->peer_uuid() << ": Peer starting";

    VLOG(2) << "RaftConfig before starting: " << consensus_->CommittedConfig().DebugString();

    RETURN_NOT_OK(consensus_->Start(bootstrap_info));
    RETURN_NOT_OK(UpdateState(TabletStatePB::BOOTSTRAPPING, TabletStatePB::RUNNING,
                              "Incorrect state to start TabletPeer, "));
  }
  // The context tracks that the current caller does not hold the lock for consensus state.
  // So mark dirty callback, e.g., consensus->ConsensusState() for master consensus callback of
  // SysCatalogStateChanged, can get the lock when needed.
  auto context =
      std::make_shared<StateChangeContext>(StateChangeReason::TABLET_PEER_STARTED, false);
  // Because we changed the tablet state, we need to re-report the tablet to the master.
  mark_dirty_clbk_.Run(context);

  return Status::OK();
}
```

And in `ConsensusServiceImpl::UpdateConsensus` we call `LookupTabletPeerOrRespond`, which checks that the state is `RUNNING`.

However, there is one more possibility for how `SnoozeFailureDetectorUnlocked` could fail. The peer could become leader while processing an `UpdateReplica`, and in that case we unregister `kTimerId` ("election-timer") from the failure detector.

```c++
Status RaftConsensus::EnsureFailureDetectorDisabledUnlocked() {
  if (PREDICT_FALSE(!FLAGS_enable_leader_failure_detection)) {
    return Status::OK();
  }

  if (!failure_detector_->IsTracking(kTimerId)) {
    return Status::OK();
  }
  return failure_detector_->UnTrack(kTimerId);
}
```

**Comment by bmatican** [comment]
@mbautin can we close this out?

**Comment by bmatican** [comment]
I believe 0d07add220f536fcf3b27402ea0086776198fa79 fixed this.
